servers/rendering/renderer_viewport.cpp

```diff
@@ -286,11 +288,17 @@ void RendererViewport::_draw_viewport(Viewport *p_viewport) {
 				// The scene renderer will still copy over the last frame, so we need to clear the render target.
 				force_clear_render_target = true;
 			}
+
+			// Calculate the frames needed for volumetric fog to converge
+			if (p_viewport->size != p_viewport->old_size && can_draw_3d) {
+				p_viewport->old_size = p_viewport->size;
+				float temporal_amount = RSG::scene->environment_get_volumetric_fog_temporal_reprojection_amount(environment);
+				float convergence_threshold = 0.01f;
+				p_viewport->frames_needed = Math::ceil(Math::log(convergence_threshold) / Math::log(temporal_amount));
```

Comment:
Maybe this is accounted elsewhere, but shouldn't this value be limited to a reasonable number of frames, so that very extreme amounts won't freeze the editor?

Reply:
The upper threshold you can set in the volumetric fog temporal reprojection amount is `0.999`, which amounts to 459 frames with this algorithm. This is indeed quite a lot if your editor is limtied to 60 Hz due to V-Sync (even worse if your PC can't keep up 60 FPS), as it means the editor will continuously redraw for almost 10 seconds every time you do something.

I'd probably cap this value to 100 frames or something.

Reply:
We could change the convergence threshold, but there's sacrifices to it.

`Convergence Threshold = 0.01`
Slow convergence, requiring more frames, but looks smoother between frames.

| Temporal Reprojection Amount | Frames Needed (Uncapped) | Frames Needed (Capped at 100) |
| :----------------: | :------: | :----: |
| 0.90  |   44    |   44   |
| 0.95  |   90    |   90   |
| 0.97  |  152   |  100  |
| 0.98  |  228   |  100  |
| 0.99  |  459   |  100  |

`Convergence Threshold = 0.1`
Faster convergence, requiring less frames, but frames will have a noticeable effect.

| Temporal Reprojection Amount | Frames Needed (Uncapped) | Frames Needed (Capped at 100) |
| :----------------: | :------: | :----: |
| 0.90  |   22    |   22   |
| 0.95  |   45    |  45   |
| 0.97  |   76    |  76  |
| 0.98  |  114   |  100  |
| 0.99  |  230   |  100  |